### PR TITLE
[FLINK-22857][table] Support evaluating expressions in SpecializedFunction

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This enables to evaluate arbitrary expressions in `SpecializedFunction`. It allows to access all built-in functions including code generated ones during runtime. It enables writing UDFs in pure Java code without adjusting the code generation at various locations. As a proof-of-concept and reference implementation this PR also adds the `ARRAY_CONTAINS` function to both Table API and SQL.

## Brief change log

- Various refactorings around context instances
- New APIs for `SpecializedFunction`
- ARRAY_CONTAINS for Table API and SQL

## Verifying this change

This change added tests and can be verified as follows: `FunctionITCase`, `CollectionFunctionsITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala [hunk 1]
```diff
@@ -396,4 +405,183 @@ object BridgingFunctionGenUtil {
...
+        createEvaluatorOrError(expression, outputDataType, args)
+      } catch {
+        case t: Throwable =>
+          throw new TableException(
+            s"Unable to create an expression evaluator for expression: $expression",
+            t)
+      }
+    }
+
+    private def createEvaluatorOrError(
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala:457]: Would be great to have some sample code in comments here, similarly to what we have for the Casting functions.
Maybe refer to the example in `FunctionITCase` with the `RowEqualityScalarFunction` custom function?

**Review by matriv — COMMENTED** [review]
Thank you @twalthr, nice stuff!
Left some comments, mostly around testing.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayContainsFunction.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    public @Nullable Boolean eval(ArrayData haystack, Object needle) {
+        try {
+            if (haystack == null) {
+                return null;
+            }
+            final int size = haystack.size();
+            for (int pos = 0; pos < size; pos++) {
+                final Object element = elementGetter.getElementOrNull(haystack, pos);
+                // handle nullability before to avoid SQL three-value logic for equality
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayContainsFunction.java:70]: I added a comment.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala [hunk 3]
```diff
@@ -351,6 +351,15 @@ class ExprCodeGenerator(ctx: CodeGeneratorContext, nullableInput: Boolean)
   }
 
   override def visitInputRef(inputRef: RexInputRef): GeneratedExpression = {
+    // for specific custom code generation
+    if (input1Type == null) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala:355]: This is a bit hacky but safe enough. The code generator is very generic with a couple of corner cases for various purposes.

**Comment by twalthr** [comment]
Thanks for the review @matriv. I addressed your comments.

**Review by matriv — APPROVED** [review]
LGTM. Thank you!

**Pull request closed without merging**